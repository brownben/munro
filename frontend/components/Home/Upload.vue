<template>
  <aside class="relative bg-main-800 text-main-600">
    <svg aria-hidden="true" class="absolute inset-0 z-0 h-full w-full">
      <defs>
        <pattern
          id=":Rb6:"
          width="90"
          height="30"
          patternUnits="userSpaceOnUse"
          x="50%"
          patternTransform="translate(0 80)"
        >
          <path d="M0 128V.5H128" fill="none" stroke="currentColor"></path>
        </pattern>
      </defs>
      <rect width="100%" height="100%" fill="url(#:Rb6:)"></rect>
    </svg>
    <div
      class="relative z-10 mx-auto max-w-screen-lg gap-8 px-6 py-10 sm:py-14 lg:px-8"
    >
      <h2
        class="text-4xl font-black leading-none tracking-tight text-white sm:text-5xl"
      >
        Stop fighting spreadsheets
      </h2>
      <p class="mt-4 text-lg leading-snug text-main-50">
        Upload results in seconds. No fuss, no hassle, just works.
      </p>
      <NuxtLink
        to="/upload/instructions"
        class="mt-8 inline-block font-medium text-white"
      >
        See how simple it is â†’
      </NuxtLink>
    </div>
  </aside>
</template>
