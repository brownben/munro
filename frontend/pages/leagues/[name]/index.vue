<script setup lang="ts">
import {
  UserIcon,
  CalculatorIcon,
  ChartBarSquareIcon,
  GlobeEuropeAfricaIcon,
} from '@heroicons/vue/24/outline/index.js'
import type { LeagueOverview } from '~/api-types'

const route = useRoute()
const loggedIn = useLoggedIn()

const { data: league, refresh } =
  await useRestrictedDataIfLoggedIn<LeagueOverview>(
    `leagues/${route.params.name}`,
    `leagues/${route.params.name}/uploadKey`
  )

const events = league.value?.events ?? []
const latestEvents = events.filter((event) => event.results_uploaded).reverse()

const todaysDate = getCurrentDate()
const nextEvent = events.filter(
  (event) => event.date >= todaysDate && !event.results_uploaded
)[0]

const scoringMethods: Record<string, [string, string]> = {
  position: ['Position Based Scoring', 'System (100 Max)'],
  position50: ['Position Based Scoring', 'System (50 Max)'],
  position99: ['Position Based Scoring', 'System (99 Max)'],
  position99average: [
    'Position Based Scoring',
    'System (99 Max, Reduced in a Draw)',
  ],
  positionDouble: ['Position Based Scoring', 'System (100 Max, Double Points)'],
  position50Double: [
    'Position Based Scoring',
    'System (50 Max, Double Points)',
  ],
  timeAverage: ['Time Relative to Average Scoring', 'System (1000 Average)'],
  timeAverage100: ['Time Relative to Average Scoring', 'System (100 Average)'],
  timeTop3: ['Time Relative to the Average Time of the Top 3', ''],
  timeTop3Adjusted: [
    'Time Relative to the Average Time of the Top 3',
    '(Adjusted by Course/Age Class)',
  ],
  timeTopAdjustedWelsh: [
    'Time Relative to Winner',
    'System (Adjusted by Course/Age Class - Welsh Multipliers)',
  ],
  positionStaggered: ['Position Based Scoring', 'System (Staggered, 60 Max)'],
  file: ['Points from File', ''],
}

if (league.value) {
  useTitle({
    title: league.value.name,
    description: `${league.value.tagline} - All the results and event information for the ${league.value.name} league on Munro`,
  })
}
</script>
<template>
  <div v-if="league">
    <Heading :title="league.name" :tagline="league.tagline">
      <ImageRow v-if="league.coordinator" :icon="UserIcon">
        Coordinated by
        <strong class="font-medium">{{ league.coordinator }}</strong>
      </ImageRow>
      <ImageRow v-if="league.scoring_method" :icon="CalculatorIcon">
        <strong class="font-medium">
          {{ scoringMethods[league.scoring_method][0] }}
        </strong>
        {{ scoringMethods[league.scoring_method][1] }}
      </ImageRow>
      <ImageRow :icon="ChartBarSquareIcon">
        Your
        <strong class="font-medium">
          best {{ league.number_of_counting_events }} scores
        </strong>
        from all
        <strong class="font-medium">
          {{ league.events.length }} events count
        </strong>
      </ImageRow>
      <ImageRow v-if="league.website" :icon="GlobeEuropeAfricaIcon" hover>
        <a :href="league.website" rel="noopener noreferrer" target="_blank">
          {{ league.website }}
        </a>
      </ImageRow>
      <template v-if="league.more_information" #extra>
        <p
          v-for="paragraph in league.more_information.split('|')"
          :key="paragraph"
          class="mt-1 text-gray-500"
        >
          {{ paragraph }}
        </p>
      </template>
    </Heading>
    <main>
      <LinksSection
        v-if="loggedIn"
        :links="[
          {
            text: 'Edit League',
            location: `/leagues/${league.name}/edit`,
          },
          {
            text: 'Create Event',
            location: `/events/create?league=${league.name}`,
          },
          {
            text: 'Manage Competitors',
            location: `/competitor-pools/${league.competitor_pool}`,
          },
        ]"
        dark
      >
        Manage your league
      </LinksSection>

      <section
        v-if="events.length >= 2 && latestEvents.length > 0"
        class="mx-auto hidden max-w-screen-lg grid-cols-1 gap-8 px-6 py-8 sm:grid sm:grid-cols-2 sm:pt-10 sm:pb-6 lg:px-8"
      >
        <div>
          <h2 class="font-bold uppercase text-gray-500 sm:pb-2">
            Latest Results
          </h2>
          <Event :event="latestEvents[0]" class="pt-4 pb-6" />
        </div>
        <div v-if="nextEvent">
          <h2 class="font-bold uppercase text-gray-500 sm:pb-2">Next Event</h2>
          <Event :event="nextEvent" class="pt-4 pb-6" />
        </div>
        <div v-else>
          <p
            class="hidden font-bold uppercase text-gray-500 sm:block sm:pb-2"
            aria-hidden="true"
          >
            &nbsp;
          </p>
          <Event
            v-if="latestEvents.length > 1"
            :event="latestEvents[1]"
            class="pt-4 pb-6"
          />
        </div>
      </section>

      <LinksSection
        v-if="league.classes.length > 0"
        :links="
          league.classes.map((course) => ({
            text: course.name,
            location: `/leagues/${league?.name}/results/${course.name}`,
          }))
        "
      >
        League Results
      </LinksSection>

      <section v-if="loggedIn">
        <div
          class="mx-auto flex max-w-screen-lg grid-cols-3 flex-col gap-8 px-6 py-6 sm:grid sm:pt-10 sm:pb-6 lg:px-8"
        >
          <div class="pb-4">
            <h2 class="col-span-1 text-2xl font-bold text-gray-500">Classes</h2>
            <NuxtLink
              :to="`/leagues/${route.params.name}/classes/create`"
              class="mt-2 inline-block rounded bg-gray-100 px-2 py-1 text-sm font-medium text-gray-700 transition hover:bg-gray-200"
            >
              + Add Class
            </NuxtLink>
          </div>

          <div class="col-span-2 -mt-6 flex flex-col divide-y">
            <LeagueClass
              v-for="cls in league.classes"
              :key="cls.name"
              :cls="cls"
              class="py-6 sm:px-4"
              @class-changed="refresh"
            />
          </div>
        </div>
      </section>

      <section
        class="mx-auto max-w-screen-lg gap-8 px-6 py-6 sm:py-12 md:py-14 lg:px-8"
      >
        <div class="flex grid-cols-3 flex-col sm:grid">
          <div class="col-span-1 pb-8 sm:pb-0">
            <h2 class="text-2xl font-bold text-gray-500">Events</h2>
            <NuxtLink
              v-if="loggedIn"
              :to="`/events/create?league=${league.name}`"
              class="mt-2 inline-block rounded bg-gray-100 px-2 py-1 text-sm font-medium text-gray-700 transition hover:bg-gray-200"
            >
              + Add Event
            </NuxtLink>
          </div>

          <div class="col-span-2 -mt-10 -mb-4 flex flex-col divide-y">
            <Event
              v-for="event in league.events"
              :key="event.id"
              :event="event"
              class="py-10 sm:px-4"
              :admin="loggedIn"
              @event-changed="refresh"
            />
          </div>
        </div>
      </section>
    </main>
  </div>
  <LazyError v-else />
</template>
