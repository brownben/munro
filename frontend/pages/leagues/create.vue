<script setup lang="ts">
import {
  IsValidURLParameter,
  IsValidURL,
  RequiredField,
} from '~/utils/validation'

const loggedIn = useLoggedIn()
if (!loggedIn.value) await redirect('/login')

const router = useRouter()

const { data: competitor_pools } = await useData<string[]>(`competitor-pools/`)

const form = reactive({
  name: '',
  tagline: '',
  year: new Date().getUTCFullYear(),
  coordinator: '',
  website: '',
  more_information: '',

  visible: true,
  scoring_method: 'position',
  number_of_counting_events: 1,
  competitor_pool: '',
})

const action = async () => {
  try {
    await usePost(`leagues/`, toRaw(form))
    await router.replace(`/leagues/${form.name}`)
  } catch (error: any) {
    if (error.data && typeof error.data?.detail === 'string')
      throw error.data.detail
    else throw 'Problem creating league, please try again.'
  }
}

useTitle({
  title: 'Create League',
  description: 'Add a new league to Munro',
})
</script>

<template>
  <div>
    <Heading title="Create League" />

    <Form button="Create" :action="action">
      <FormHeading
        title="Basic Information"
        description="The identifying data for the league."
      />

      <Input
        v-model.trim="form.name"
        label="Name:"
        class="col-span-2"
        :validator="IsValidURLParameter('a name')"
      />
      <Input
        v-model.trim="form.tagline"
        label="Description:"
        class="col-span-2"
      />
      <Input
        v-model.trim="form.coordinator"
        label="Coordinator:"
        class="col-span-2"
      />
      <InputNumber
        v-model="form.year"
        label="Year:"
        :min="1990"
        :max="2050"
        class="col-span-2"
      />
      <Input
        v-model.trim="form.website"
        label="Website: (URL)"
        type="url"
        class="col-span-2"
        :validator="IsValidURL"
      />
      <InputTextarea
        v-model.trim="form.more_information"
        label="More Information:"
        class="col-span-2"
      />

      <FormHeading
        title="Scoring"
        description="How the scores should be calculated for the league"
      />

      <InputDropdown
        v-model="form.scoring_method"
        :list="[
          { value: 'position', text: 'Position Based (100 Max)' },
          { value: 'position50', text: 'Position Based (50 Max)' },
          { value: 'position99', text: 'Position Based (99 Max)' },
          {
            value: 'position99average',
            text: 'Position Based (99 Max, Reduced in a Draw)',
          },
          {
            value: 'positionDouble',
            text: 'Position Based (100 Max, Double Points)',
          },
          {
            value: 'position50Double',
            text: 'Position Based (50 Max, Double Points)',
          },
          {
            value: 'positionStaggered',
            text: 'Position Based (Staggered, 60 Max)',
          },
          {
            value: 'timeAverage',
            text: 'Relative to Average Time (1000 Average)',
          },
          {
            value: 'timeAverage100',
            text: 'Relative to Average Time (100 Average)',
          },
          { value: 'timeTop3', text: 'Relative to Top 3 Times' },
          {
            value: 'timeTopAdjustedWelsh',
            text: 'Relative to Winner\'s Time (Welsh Multipliers)',
          },
          { value: 'file', text: 'From Upload File' },
        ]"
        label="Scoring Method:"
        :validator="RequiredField('a scoring method', true)"
        class="col-span-2"
      />
      <InputNumber
        v-model="form.number_of_counting_events"
        :validator="RequiredField('a number of counting events')"
        :min="1"
        label="Maximum Number of Counting Events:"
        class="col-span-2"
      />

      <FormHeading
        title="Additional Settings"
        description="Adjust advanced settings and enable additional features"
      />
      <InputDropdown
        v-model="form.competitor_pool"
        :list="competitor_pools?.map((name) => ({ value: name, text: name }))"
        label="Competitor Pool:"
        include-blank
        class="col-span-2"
      />
    </Form>
  </div>
</template>
