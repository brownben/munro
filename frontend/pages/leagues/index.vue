<script setup lang="ts">
import type { League } from '~/api-types'

const { data: leagues } = await useData<League[]>('leagues/')

useTitle({
  title: 'Leagues',
  description: 'All the leagues with results on Munro',
})
</script>

<template>
  <div>
    <Heading title="Leagues" />
    <main
      class="mx-auto max-w-screen-lg gap-8 px-6 py-10 sm:py-12 md:py-14 lg:px-8"
    >
      <div
        class="col-span-2 -mx-4 -mb-4 -mt-8 flex flex-col divide-y dark:divide-gray-600"
      >
        <League
          v-for="league in leagues"
          :key="league.name"
          :league="league"
          class="px-4 py-10"
        />
      </div>
    </main>
  </div>
</template>
