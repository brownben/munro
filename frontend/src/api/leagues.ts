import { getData, postData, putData, deleteData } from './requests'

export const getLeague = (name: string): Promise<League | null> =>
  getData<League | null>({
    apiLocation: `/api/leagues/${name}`,
    customErrorMessage: 'Problem Fetching League Details',
  })

export const getLeagues = (): Promise<League[] | null> =>
  getData<League[]>({
    apiLocation: '/api/leagues',
    customErrorMessage: 'Problem Fetching Leagues',
  })

export const createLeague = (data: LeagueForm): Promise<ServerMessage | null> =>
  postData<ServerMessage>({
    apiLocation: `/api/leagues`,
    data,
    customErrorMessage: 'Problem Creating League',
    customSuccessMessage: `League \`${data.name}\` Created`,
    customErrorHandler: true,
  })

export const updateLeague = (data: LeagueForm): Promise<ServerMessage | null> =>
  putData<ServerMessage>({
    apiLocation: `/api/leagues/${data.name}`,
    data,
    customErrorMessage: 'Problem Updating League',
    customSuccessMessage: `League \`${data.name}\` Updated`,
    customErrorHandler: true,
  })

export const deleteLeague = (name: string): Promise<ServerMessage | null> =>
  deleteData<ServerMessage>({
    apiLocation: `/api/leagues/${name}`,
    customErrorMessage: 'Problem Deleting League',
    customErrorHandler: true,
  })
