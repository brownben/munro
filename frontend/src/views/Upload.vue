<!--
  Upload Page

  Upload results to database.
-->

<template>
  <div>
    <h1>Upload Results</h1>
    <label>Event ID:</label>
    <input v-model.trim.lazy="eventId" type="text" @change="findEvent()">
    <p v-show="event.name" id="eventName">
      <b>Event Name:</b>
      {{ event.name }}
    </p>
    <label>Upload Key:</label>
    <input v-model.trim="uploadKey" type="text">
    <label>Results File:</label>
    <div class="file-input">
      <label for="file" class="button">Browse for File</label>
      <input id="file" type="file" accept=".csv" @change="fileChange">
      <p>{{ fileName }}</p>
    </div>
    <!-- If Event already have results, confirm they want to overwrite -->
    <Checkbox v-if="event.resultUploaded" v-model="overwrite" label="Overwrite Existing Results:" />
    <!-- Only show upload once all fields have been filled -->
    <button v-if="eventId && uploadKey && file" id="uploadButton" @click="uploadFile">Upload File</button>
  </div>
</template>

<script>
import axios from 'axios'
import CheckboxInput from '@/components/CheckboxInput'

export default {
  components: {
    'Checkbox': CheckboxInput,
  },

  data: function () {
    return {
      eventId: '',
      uploadKey: '',
      event: {},
      fileName: 'Select a File',
      file: '',
      overwrite: false,
    }
  },

  // On load
  mounted: function () {
    // If passed with Event ID, autofill Event ID
    if (this.$route.params.id) {
      this.eventId = this.$route.params.id
      this.findEvent()
    }
  },

  methods: {
    findEvent: function () {
      // Fetch event details so name of event can be checked and if results are uploaded
      return axios.get('/api/events/' + this.eventId)
        .then(response => {
          this.event = response.data
          if (!this.event.name) {
            this.event = {}
            this.event.name = 'No Event Found'
          }
        })
        .catch(() => this.$messages.addMessage('Problem Fetching Event Name'))
    },

    fileChange (event) {
      // When the file selected has been changed
      const storage = event.target || event.dataTransfer
      const files = storage.files
      if (!files || !files.length) return false
      this.fileName = files[0].name
      this.readFile(files[0])
    },

    readFile (file) {
      // read file using FileReader and save in data
      const reader = new FileReader()
      reader.onload = event => { this.file = event.target.result }
      reader.readAsText(file)
    },

    uploadFile () {
      // Send data to the server
      return axios.post('/api/upload', {
        eventId: this.eventId,
        uploadKey: this.uploadKey,
        file: this.file,
        overwrite: this.overwrite,
      })
        .then(() => {
          this.$messages.addMessage('Results Uploaded Successfully')
          this.$router.push('/leagues/' + this.event.league)
        })
        .catch(error => this.$messages.addMessage(error.response.data.message))
    },
  },
}
</script>

<style lang="stylus" scoped>
@import '../assets/styles/helpers.styl'
@import '../assets/styles/inputs.styl'

#router-view
  margin-left: 5%
  padding-top: 1rem
  width: 90%

h1
  margin-bottom: 0.5rem

button
  margin: 0.5rem 0

#eventName
  margin: 0 0 0.5rem

input[type='file']
  display: none

.file-input
  p, label
    display: inline-block
    margin-top: 0
    margin-right: 0.5rem
    font-size: 1rem

input, select
  width: 100% !important

#uploadButton
  margin-top: 0.5rem
</style>
