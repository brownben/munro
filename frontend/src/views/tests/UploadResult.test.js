import axios from 'axios'

import { mount } from '@vue/test-utils'

import Upload from '@/views/UploadResult.vue'
import { sampleSingleLeague } from '@/tests/test data/leagues'

jest.mock('axios')
const flushPromises = () =>
  new Promise((resolve) => global.setImmediate(resolve))

beforeEach(() => {
  jest.resetAllMocks()
  axios.get.mockRejectedValue()
})

test('Is a Vue Instance', () => {
  const wrapper = mount(Upload, {
    mocks: {
      $route: { params: { id: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly', () => {
  const wrapper = mount(Upload, {
    mocks: {
      $route: { params: { id: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Upload Buttons Shows Correctly', () => {
  const wrapper = mount(Upload, {
    mocks: {
      $route: { params: { id: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({
    eventId: '1',
    name: '2',
    course: '3',
    time: '2',
  })
  expect(wrapper.findAll('button').length).toBe(1)
  expect(wrapper.element).toMatchSnapshot()
  wrapper.setData({
    eventId: '1',
    name: '2',
    course: '',
    time: '2',
  })
  expect(wrapper.findAll('button').length).toBe(0)
  wrapper.setData({
    eventId: '',
    name: '2',
    course: '',
    time: '',
  })
  expect(wrapper.findAll('button').length).toBe(0)
})

test('Get Events - Correct API Call', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = mount(Upload, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: 'Test' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => {},
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue()
  await wrapper.vm.getEvents()
  expect(axios.get).toHaveBeenCalledTimes(1)
  expect(axios.get).toHaveBeenLastCalledWith('/api/events')
})

test('Get Events - Success', async () => {
  const mockAddMessage = jest.fn()

  const wrapper = mount(Upload, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => {},
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue({
    data: [
      { userSubmittedResults: false, name: 'a' },
      { userSubmittedResults: true, name: 'b' },
      { userSubmittedResults: true, name: 'c' },
    ],
  })
  await wrapper.vm.getEvents()
  expect(wrapper.vm.events).toEqual([
    { userSubmittedResults: true, name: 'b' },
    { userSubmittedResults: true, name: 'c' },
  ])
})

test('Get Events - No Events', async () => {
  const mockAddMessage = jest.fn()

  const wrapper = mount(Upload, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => {},
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue({
    data: [],
  })
  await wrapper.vm.getEvents()
  expect(wrapper.vm.events).toEqual([])
  expect(mockAddMessage).toHaveBeenLastCalledWith(
    'Sorry, No Events Found to Submit Results For'
  )
})

test('Get Events - Error', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = mount(Upload, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => {},
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockRejectedValue()
  await wrapper.vm.getEvents()
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith(
    'Problem Getting Event Details'
  )
})

test('Get Leagues - Request Called Correctly', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(Upload, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.resetAllMocks()
  axios.get.mockResolvedValue({ data: sampleSingleLeague })
  await wrapper.vm.getLeagues()
  expect(axios.get).toHaveBeenCalledTimes(1)
  expect(axios.get).toHaveBeenLastCalledWith('/api/leagues')
})

test('Get Leagues - Processes Response Correctly', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(Upload, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  axios.get.mockResolvedValue({ data: [{ league: '1' }, { league: '2' }] })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([{ league: '1' }, { league: '2' }])

  axios.get.mockResolvedValue({ data: [] })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([])

  axios.get.mockResolvedValue({
    data: [
      { league: 'a', property: 3 },
      { league: '2', another: 4 },
    ],
  })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([
    { league: 'a', property: 3 },
    { league: '2', another: 4 },
  ])
})

test('Get Leagues - Shows Message on Error', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(Upload, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  axios.get.mockRejectedValue()
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([])
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Problem Fetching League Details'
  )
})

test('Upload Result - Success', async () => {
  const mockAddMessage = jest.fn()
  const mockRouterPush = jest.fn()
  const wrapper = mount(Upload, {
    mocks: {
      $route: { params: { id: '' } },
      $router: { push: mockRouterPush },
      $messages: { addMessage: mockAddMessage },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({
    eventId: '1',
    course: '2',
    name: '3',
    time: '4',
  })
  axios.post.mockResolvedValue()
  await wrapper.vm.uploadResult()
  expect(axios.post).toHaveBeenCalledTimes(1)
  expect(axios.post).toHaveBeenLastCalledWith('/api/upload/result', {
    eventId: '1',
    course: '2',
    name: '3',
    time: '4',
  })
  expect(mockRouterPush).toHaveBeenCalledTimes(1)
  expect(mockRouterPush).toHaveBeenLastCalledWith('/events/1/results')
  expect(mockAddMessage).toHaveBeenCalledWith('Result Uploaded Successfully')
})

test('Upload Result - Failure', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = mount(Upload, {
    mocks: {
      $route: { params: { id: '' } },
      $messages: { addMessage: mockAddMessage },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  axios.post.mockRejectedValue({ response: { data: { message: 'Error' } } })
  await wrapper.vm.uploadResult()
  expect(mockAddMessage).toHaveBeenLastCalledWith('Error')
})
