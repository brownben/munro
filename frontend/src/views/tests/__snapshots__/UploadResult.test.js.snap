// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly 1`] = `
<div
  class="flex flex-col flex-grow h-full"
>
  <header
    class="text-gray-900 bg-white"
  >
    <div
      class="max-w-screen-xl px-6 pb-6 mx-auto lg:px-8 pt-6"
    >
      <h1
        class="text-3xl font-bold leading-tight font-heading"
      >
        
          Submit Result
        
      </h1>
    </div>
  </header>
   
  <main
    class="flex flex-col flex-grow"
  >
    <!---->
     
    <!---->
     
    <div
      class="flex-grow bg-white"
    >
      <div
        class="max-w-screen-xl pb-8 mx-auto sm:pb-10 xl:pb-12 lg:px-8 sm:px-6"
      >
        <div
          class="grid grid-cols-2 gap-6 px-6 sm:px-0 sm:gap-8"
        >
          <vue-headful-stub
            description="Manually Submit a Result to Munro - League Results. Sorted. Sports League Results Calculated Quick and Easily, with Results Sorting and Filtering Options"
            head="[object Object]"
            title="Munro - Submit Result"
            url="https://munro-leagues.herokuapp.com/upload/result"
          />
           
          <div
            class="col-span-2"
          >
            <div>
              <label
                class="block pb-1 select-none font-heading text-gray-600"
                for="Event:"
              >
                
    Event:
  
              </label>
               
              <div
                class="relative"
              >
                <span
                  class="absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none"
                >
                  <svg
                    class="w-5 h-5 text-gray-400"
                    fill="none"
                    stroke="currentColor"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 9l-7 7-7-7"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="1.5"
                    />
                  </svg>
                </span>
                 
                <select
                  class="w-full px-3 py-2 font-sans text-gray-900 transition duration-300 ease-in-out bg-white border outline-none appearance-none rounded-shape focus:shadow-outline focus:border-main-400"
                  id="Event:"
                >
                  <option />
                   
                </select>
              </div>
            </div>
             
            <div
              class="mt-4"
            >
              <label
                class="block pb-1 select-none font-heading text-gray-600"
                for="Your Course:"
              >
                
    Your Course:
  
              </label>
               
              <div
                class="relative"
              >
                <span
                  class="absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none"
                >
                  <svg
                    class="w-5 h-5 text-gray-400"
                    fill="none"
                    stroke="currentColor"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 9l-7 7-7-7"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="1.5"
                    />
                  </svg>
                </span>
                 
                <select
                  class="w-full px-3 py-2 font-sans text-gray-900 transition duration-300 ease-in-out bg-white border outline-none appearance-none rounded-shape focus:shadow-outline focus:border-main-400"
                  id="Your Course:"
                >
                  <option />
                   
                </select>
              </div>
            </div>
             
            <div
              class="mt-4"
            >
              <label
                class="block pb-1 select-none font-heading text-gray-600"
                for="Your Name:"
              >
                
    Your Name:
  
              </label>
               
              <input
                class="w-full px-3 py-2 font-sans text-gray-900 transition duration-300 ease-in-out bg-white border outline-none appearance-none rounded-shape focus:shadow-outline focus:border-main-400"
                id="Your Name:"
                type="text"
              />
            </div>
             
            <div
              class="mt-4"
            >
              <label
                class="block pb-1 select-none font-heading text-gray-600"
                for="Your Time: (HH:MM:SS)"
              >
                
    Your Time: (HH:MM:SS)
  
              </label>
               
              <input
                class="w-full px-3 py-2 font-sans text-gray-900 transition duration-300 ease-in-out bg-white border outline-none appearance-none rounded-shape focus:shadow-outline focus:border-main-400"
                id="Your Time: (HH:MM:SS)"
                type="text"
              />
            </div>
             
            <!---->
          </div>
        </div>
      </div>
    </div>
  </main>
   
  <!---->
</div>
`;

exports[`Upload Buttons Shows Correctly 1`] = `
<div
  class="flex flex-col flex-grow h-full"
>
  <header
    class="text-gray-900 bg-white"
  >
    <div
      class="max-w-screen-xl px-6 pb-6 mx-auto lg:px-8 pt-6"
    >
      <h1
        class="text-3xl font-bold leading-tight font-heading"
      >
        
          Submit Result
        
      </h1>
    </div>
  </header>
   
  <main
    class="flex flex-col flex-grow"
  >
    <!---->
     
    <!---->
     
    <div
      class="flex-grow bg-white"
    >
      <div
        class="max-w-screen-xl pb-8 mx-auto sm:pb-10 xl:pb-12 lg:px-8 sm:px-6"
      >
        <div
          class="grid grid-cols-2 gap-6 px-6 sm:px-0 sm:gap-8"
        >
          <vue-headful-stub
            description="Manually Submit a Result to Munro - League Results. Sorted. Sports League Results Calculated Quick and Easily, with Results Sorting and Filtering Options"
            head="[object Object]"
            title="Munro - Submit Result"
            url="https://munro-leagues.herokuapp.com/upload/result"
          />
           
          <div
            class="col-span-2"
          >
            <div>
              <label
                class="block pb-1 select-none font-heading text-gray-600"
                for="Event:"
              >
                
    Event:
  
              </label>
               
              <div
                class="relative"
              >
                <span
                  class="absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none"
                >
                  <svg
                    class="w-5 h-5 text-gray-400"
                    fill="none"
                    stroke="currentColor"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 9l-7 7-7-7"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="1.5"
                    />
                  </svg>
                </span>
                 
                <select
                  class="w-full px-3 py-2 font-sans text-gray-900 transition duration-300 ease-in-out bg-white border outline-none appearance-none rounded-shape focus:shadow-outline focus:border-main-400"
                  id="Event:"
                >
                  <option />
                   
                </select>
              </div>
            </div>
             
            <div
              class="mt-4"
            >
              <label
                class="block pb-1 select-none font-heading text-gray-600"
                for="Your Course:"
              >
                
    Your Course:
  
              </label>
               
              <div
                class="relative"
              >
                <span
                  class="absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none"
                >
                  <svg
                    class="w-5 h-5 text-gray-400"
                    fill="none"
                    stroke="currentColor"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 9l-7 7-7-7"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="1.5"
                    />
                  </svg>
                </span>
                 
                <select
                  class="w-full px-3 py-2 font-sans text-gray-900 transition duration-300 ease-in-out bg-white border outline-none appearance-none rounded-shape focus:shadow-outline focus:border-main-400"
                  id="Your Course:"
                >
                  <option />
                   
                </select>
              </div>
            </div>
             
            <div
              class="mt-4"
            >
              <label
                class="block pb-1 select-none font-heading text-gray-600"
                for="Your Name:"
              >
                
    Your Name:
  
              </label>
               
              <input
                class="w-full px-3 py-2 font-sans text-gray-900 transition duration-300 ease-in-out bg-white border outline-none appearance-none rounded-shape focus:shadow-outline focus:border-main-400"
                id="Your Name:"
                type="text"
              />
            </div>
             
            <div
              class="mt-4"
            >
              <label
                class="block pb-1 select-none font-heading text-gray-600"
                for="Your Time: (HH:MM:SS)"
              >
                
    Your Time: (HH:MM:SS)
  
              </label>
               
              <input
                class="w-full px-3 py-2 font-sans text-gray-900 transition duration-300 ease-in-out bg-white border outline-none appearance-none rounded-shape focus:shadow-outline focus:border-main-400"
                id="Your Time: (HH:MM:SS)"
                type="text"
              />
            </div>
             
            <button
              class="mt-8 button-lg"
            >
              
      Submit Result
    
            </button>
          </div>
        </div>
      </div>
    </div>
  </main>
   
  <!---->
</div>
`;
